rolling-shutter/shmsg/eonpublickey.go

```diff
@@ -21,40 +21,51 @@ func NewSignedEonPublicKey(
 	eon uint64,
 	privKey *ecdsa.PrivateKey,
 ) (*EonPublicKey, error) {
-	pubKey := &EonPublicKey{
+	candidate := &EonPublicKeyCandidate{
 		InstanceID:        instanceID,
 		PublicKey:         eonPublicKey,
 		ActivationBlock:   activationBlock,
-		KeyperIndex:       keyperIndex,
 		KeyperConfigIndex: keyperConfigIndex,
 		Eon:               eon,
 	}
-	var err error
-
-	pubKey.Signature, err = ethcrypto.Sign(pubKey.Hash(), privKey)
-	if err != nil {
-		return nil, err
-	}
-	return pubKey, nil
+	return candidate.Sign(privKey, keyperIndex)
 }
 
 func (e *EonPublicKey) VerifySignature(address common.Address) (bool, error) {
-	pubkey, err := ethcrypto.SigToPub(e.Hash(), e.Signature)
+	pubkey, err := ethcrypto.SigToPub(e.Candidate.Hash(), e.Signature)
 	if err != nil {
 		return false, err
 	}
 	recoveredAddress := ethcrypto.PubkeyToAddress(*pubkey)
 	return recoveredAddress == address, nil
 }
 
-func (e *EonPublicKey) Hash() []byte {
+func (e *EonPublicKeyCandidate) Hash() []byte {
 	hash := sha3.New256()
 	hash.Write(eonPubKeyHashPrefix)
 	_ = binary.Write(hash, binary.BigEndian, e.InstanceID)
 	_ = binary.Write(hash, binary.BigEndian, e.ActivationBlock)
-	_ = binary.Write(hash, binary.BigEndian, e.KeyperIndex)
 	_ = binary.Write(hash, binary.BigEndian, e.KeyperConfigIndex)
 	_ = binary.Write(hash, binary.BigEndian, e.Eon)
 	hash.Write(e.PublicKey)
 	return hash.Sum(nil)
 }
+
+// Sign signs the eon public key candidate and returns an eon public key.
+func (e *EonPublicKeyCandidate) Sign(
+	privKey *ecdsa.PrivateKey, keyperIndex uint64,
+) (*EonPublicKey, error) {
+	signature, err := ethcrypto.Sign(e.Hash(), privKey)
+	if err != nil {
+		return nil, err
+	}
+	return &EonPublicKey{
+		Candidate:   e,
+		KeyperIndex: keyperIndex,
+		Signature:   signature,
+	}, nil
+}
+
```

Comment:
I see your point about the additional argument (keyper-index), which also already bugged me. Maybe we should completely drop this. In this case I would be fine with also dropping the wrapper class for consistency (DecryptionTrigger also does not have a wrapper class).

> It does not set the signature in-place on the instance but returns an instance where the signature is set

of course this is how it should be! clojure ftw!

Reply:
Another reason for doing this, is that we also do not pass the keyper index as part of our response in the collator's HTTP interface. We just pass the signatures.

Reply:
> of course this is how it should be! clojure ftw!
I agree, it's just that in geth code it is done like this so it would serve consistency.

In general I'm fine with keeping this code as is, and I'm fine with keeping the wrapper class.
If we want to change it, my preference would be to just drop the Sign method and implement everything in the NewSignedEonPublicKey constructor.